### ISSUE TITLE
SCIM integration with OpenID Connect

### ISSUE THREAD

**Issue by jainh** [coreos/dex#732]
Hi,
Does OpenID Connect implementation support the following draft [Draft](http://openid.net/specs/openid-connect-scim-profile-1_0.html) ?
Also, how to integrate SCIM within the dex and harmonize it with existing OpenID connect?

**Comment by ericchiang** [comment]
To the best of my understanding SCIM is a mechanism for asynchronous updates between user management systems. Dex isn't really a user management system, it federates to existing ones. What minimal representations of users it does have can be managed from the API.

I'm a little confused what you're specifically proposing. This seems like an orthogonal issue to what dex does (authentication vs user management).

**Comment by jainh** [comment]
@ericchiang  SCIM is definitely for user management, but it also allows to add more attributes related to user info that could not be available from standard LDAP.  However, OpenID connect supports SCIM based service for userinfo that can in turn connect to LDAP or other database.  Therefore, wondering how to provision user with dex stack, does openid connect implementation in dex already has provision for the following attributes (scim_id, scim_location, scim_profile, scim_endpoint) as per draft ? Then i need to register client with those information and  on the other side SCIM service to support that interaction.

**Comment by ericchiang** [comment]
> Therefore, wondering how to provision user with dex stack, does openid connect implementation in dex already has provision for the following attributes (scim_id, scim_location, scim_profile, scim_endpoint) as per draft?

Dex does not. It only implements the core and provider metadata specs, which are the most commonly implemented throughout providers.

Let me read over that draft and get back to you on this.

**Comment by ericchiang** [comment]
So basically you'd like to add extra fields to the discovery metadata, id token, and clients?

**Comment by ericchiang** [comment]
Do you have a specific client that uses this info I can look at? Or are you planning on rolling your own?

**Comment by jainh** [comment]
This specification defines the following metadata:
Discovery metadata indicating the availability of a SCIM protocol base endpoint.
Dynamic registration metadata that is used to indicate a clients intent to use the SCIM protocol and its associated endpoint.
An additional ID Token claim which specifies the SCIM resource endpoint and identifier associated with the authenticate subject.
In addition to the above metadata attributes and claims, the specification will also show how a client MAY access the SCIM endpoint.

**Comment by ericchiang** [comment]
Yes, I read the spec. We don't have dynamic client registration but we still have additional attributes on clients.

I'm curious what you're using this for. I'm not super familiar with the SCIM ecosystem and I'd like to understand what technologies you plan to use this feature with.

**Comment by jainh** [comment]
@ericchiang  So, the first important thing is that, instead of adding Rest to Ldap, SCIM is the standard and make easier to Restify provisioning of the users, groups and extensible for attributes that can be part of the token claim set.

Second, it may require to move identity from Active directory or Ldap to cloud and it is useful there as well. There are other feature set like password reset, migration, extensibility. It seems as an Identity management service that can interact with internal system and external as well.

**Comment by jainh** [comment]
@ericchiang can you assign  Priority to this proposal ?

**Comment by ericchiang** [comment]
It's been assigned one. We're not working on it.

For us, supporting new protocols involves learning about the providers that implement them, getting good examples of clients, understanding the security implications, etc. SCIM is a big spec and most implementations seem to be closed source so its hard for us to gauge the amount of work, ongoing maintenance, and benefit of this feature. SCIM integration isn't a huge priority for us so we've not been doing this initial investigation.

If you have specific readings on SCIM and OpenID Connect integrations or software that uses this I'd be happy to explore it a little bit more.
